presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -115,6 +118,19 @@ public Session getSession()
         return session;
     }
 
+    public long getPeakMemoryInBytes()
+    {
+        return peakMemory.get();
+    }
+
+    public void updateMemoryUsage(long deltaMemoryInBytes)
+    {
+        currentMemory.addAndGet(deltaMemoryInBytes);
+        if (currentMemory.get() > peakMemory.get()) {
+            peakMemory.getAndSet(currentMemory.get());
```

Comment:
there's a "time of check to time of use" vulnerability here, if two threads call this concurrently. For example, suppose thread 1 passes the condition of the if (`currentMemory > peakMemory`), but then the query finishes and another thread decrements `currentMemory` to zero. This will incorrectly set `peakMemory` to zero when thread 1 executes the `getAndSet`